src/opnsense/mvc/app/models/OPNsense/Base/FieldTypes/BaseField.php

```diff
@@ -472,7 +499,25 @@ public function getChild($name)
         }
     }
 
-    public function isRequired()
+    /**
+     * check if current value is empty  (either boolean field as false or an empty field)
+     * @return bool
+     */
+    public function isEmpty(): bool
+    {
+        return empty($this->getCurrentValue()); // "0" is already considered false by `empty()`
+    }
+
+    /**
+     * check if current value is NOT empty  (either boolean field as false or an empty field)
+     * @return bool
+     */
+    public function isNotEmpty(): bool
```

Comment:
FWIW, I’m not a fan of inversions such as this one being anchored as a separate function. Since the logic of isEmpty can be overridden I don’t see how isNotEmpty would not adhere to !isEmpty ever. If anything it leaves room for future mistakes. Otherwise I’m quite happy with this effort. Thanks!

Reply:
@fichtner hum :thinking: ok then I can remove `isEmpty()` and keep only `isNotEmpty()` as this is the key to use the [Nullsafe methods and properties](https://www.php.net/manual/en/language.oop5.basic.php#language.oop5.basic.nullsafe) from PHP 8.0.0 and dealing with `null` value at the same time. eg like in

https://github.com/opnsense/core/pull/7879/files#diff-83b1386f39d4edc3287f6c60e97b8311c848e33e80b8bea0b2c006e5689133eeR56-R59


---
Sample test code for the logic
```php
<?php

class Test {
  var $foo=null;
}

class Test2 {
  var $bar=null;
  function isNotEmpty() {
    return !empty(trim((string)$this->bar));
  }
}

function dotest($f) {
  #var_dump($f->foo?->bar);
  #var_dump($f->foo?->isNotEmpty());

  if ($f->foo?->isNotEmpty()) {
    return 'countNotEmpty';
  }
  return 'countEmpty';
}

$t = new Test();
$t->foo = null;
var_dump(assert(dotest($t) == 'countEmpty'));

# expect: countEmpty
$t->foo = new Test2();
var_dump(assert(dotest($t) == 'countEmpty'));

# expect: countEmpty
$t->foo->bar = "";
var_dump(assert(dotest($t) == 'countEmpty'));

# expect: countEmpty
$t->foo->bar = "   ";
var_dump(assert(dotest($t) == 'countEmpty'));

# expect: countNotEmpty
$t->foo->bar = "yallah";
var_dump(assert(dotest($t) == 'countNotEmpty'));


if (null) {
  throw new Exception("SHOULD NOT HAPPEN");
}
```

Reply:
New commit pushed for the `isEmpty()` removal

Reply:
I’m sorry for causing this change here. We would prefer to keep the isEmpty because !isNotEmpty is going to be much harder to read.

Reply:
> much harder to read

agree that was why I defined both. moving to `isEmpty()` would imply to change 
```php
if ($parentNode->$fieldname?->isNotEmpty()) {
    $countNotEmpty++;
} else {
    $countEmpty++;
}
```
to something like
```php
if (is_null($parentNode->$fieldname) || $parentNode->$fieldname->isEmpty()) {
    $countEmpty++;
} else {
    $countNotEmpty++;
}
```

A bit longer, but probably easier to read for everyone :smirk: 
go for that?

Reply:
@fichtner done with latest commit...

(we should not forget to squash all commits once we all agree on the code)

Reply:
I think we never used the `?->` operator before and there are certainly some risk involved in using that in conditional statements